File: src/test/java/no/nav/veilarbvedtaksstotte/utils/TestUtils.java
package no.nav.veilarbvedtaksstotte.utils;

import lombok.SneakyThrows;

import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.concurrent.TimeUnit;

public class TestUtils {

    private TestUtils(){}

    public static final String KAFKA_IMAGE = "confluentinc/cp-kafka:5.4.3";

    @SneakyThrows
    public static String readTestResourceFile(String fileName) {
        URL fileUrl = TestUtils.class.getClassLoader().getResource(fileName);
        Path resPath = Paths.get(fileUrl.toURI());
        return new String(Files.readAllBytes(resPath), StandardCharsets.UTF_8);
    }

    @SneakyThrows
    public static void verifiserAsynkront(long timeout, TimeUnit unit, Runnable verifiser) {
        long timeoutMillis = unit.toMillis(timeout);
        boolean prosessert = false;
        boolean timedOut = false;
        long start = System.currentTimeMillis();
        while (!prosessert) {
            try {
                Thread.sleep(10);
                long current = System.currentTimeMillis();
                timedOut = current - start > timeoutMillis;
                verifiser.run();
                prosessert = true;
            } catch (Throwable a) {
                if (timedOut) {
                    throw a;
                }
            }
        }
    }

}

File: src/test/java/no/nav/veilarbvedtaksstotte/utils/TestUtils.kt
package no.nav.veilarbvedtaksstotte.utils

import com.github.tomakehurst.wiremock.client.WireMock
import com.github.tomakehurst.wiremock.matching.StringValuePattern
import org.junit.Assert
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Paths
import java.util.concurrent.TimeUnit

object TestUtils {

    @JvmStatic
    fun readTestResourceFile(fileName: String?): String {
        val fileUrl = TestUtils::class.java.classLoader.getResource(fileName)
        val resPath = Paths.get(fileUrl.toURI())
        return String(Files.readAllBytes(resPath), StandardCharsets.UTF_8)
    }

    @JvmStatic
    fun givenWiremockOkJsonResponse(url: String, json: String) {
        WireMock.givenThat(
            WireMock.get(WireMock.urlEqualTo(url))
                .willReturn(WireMock.aResponse()
                    .withStatus(200)
                    .withHeader("Content-Type", "application/json")
                    .withBody(json)
                )
        )
    }

    @JvmStatic
    fun givenWiremockOkJsonResponseForPost(url: String, requestBody: StringValuePattern, json: String) {
        WireMock.givenThat(
            WireMock.post(url)
                .withRequestBody(requestBody)
                .willReturn(WireMock.aResponse()
                    .withStatus(200)
                    .withHeader("Content-Type", "application/json")
                    .withBody(json)
                )
        )
    }

    fun verifiserAsynkront(timeout: Long, unit: TimeUnit, verifiser: Runnable) {
        val timeoutMillis = unit.toMillis(timeout)
        var prosessert = false
        var timedOut = false
        val start = System.currentTimeMillis()
        while (!prosessert) {
            try {
                Thread.sleep(10)
                val current = System.currentTimeMillis()
                timedOut = current - start > timeoutMillis
                verifiser.run()
                prosessert = true
            } catch (a: Throwable) {
                if (timedOut) {
                    throw a
                }
            }
        }
    }

    @JvmStatic
    fun randomAlphabetic(length: Int): String =
        (1..length)
            .map { ('A'..'Z') + ('a'..'z') }
            .flatten()
            .shuffled()
            .take(length)
            .joinToString("")

    @JvmStatic
    fun randomNumeric(length: Int): String =
        (1..length)
            .map { ('0'..'9').random() }
            .joinToString("")

    inline fun <reified T : Throwable> assertThrowsWithMessage(messsage: String, runnable: Runnable) {
        val exception = Assert.assertThrows(
            T::class.java
        ) {
            runnable.run()
        }
        Assert.assertEquals(messsage, exception.message)
    }
}
